pkg/cmd/server/start/start_master.go

```diff
@@ -513,20 +515,21 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
 	oc.RunSecurityAllocationController()
 
 	if kc != nil {
+		// ensure pending namespace deletions are not blocked by later controller startup failures
+		kc.RunNamespaceController()
+		// called by admission control
+		kc.RunResourceQuotaManager()
+
 		_, rcClient, err := oc.GetServiceAccountClients(oc.ReplicationControllerServiceAccount)
```

Comment:
Anything that is like an "ensure" that fails should be done as early as
possible and terminate the controller process if it can't make progress.
The core controllers loop in kube master is an example of that.

On Fri, Aug 21, 2015 at 4:17 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/cmd/server/start/start_master.go
> <https://github.com/openshift/origin/pull/4318#discussion_r37673784>:
>
> > @@ -513,20 +515,21 @@ func startControllers(oc *origin.MasterConfig, kc *kubernetes.MasterConfig) erro
> >  	oc.RunSecurityAllocationController()
> >
> >  	if kc != nil {
> > +		// ensure pending namespace deletions are not blocked by later controller startup failures
> > +		kc.RunNamespaceController()
> > +		// called by admission control
> > +		kc.RunResourceQuotaManager()
> > +
> >  		_, rcClient, err := oc.GetServiceAccountClients(oc.ReplicationControllerServiceAccount)
>
> In the openshift-infra deletion case, if there is no token, no token can
> be created until the namespace deletion completes, and there is actually
> nothing this controller can do to work successfully until the deletion
> completes, and another server start recreates it. I think a more
> fundamental question is what controllers should do if they die/fail...
> restart (maybe forever), die and leave other controllers running (which is
> what some may do today), or take down the controller process.
> @smarterclayton <https://github.com/smarterclayton>, opinions on that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4318/files#r37673784>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift